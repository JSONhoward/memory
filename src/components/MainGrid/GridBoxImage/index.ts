import GridBoxImage from './GridBoxImage'
export default GridBoxImage