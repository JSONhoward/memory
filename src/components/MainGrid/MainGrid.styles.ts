import styled from 'styled-components';

export const StyledMainGrid = styled('div') <{ size: number }>`
display: grid;
grid-template-columns: ${props => {
        switch (props.size) {
            case 12:
                return `repeat(4, 1fr)`;
            case 20:
                return `repeat(5, 1fr)`;
            case 30:
                return `repeat(6, 1fr)`;
            default:
                return `repeat(5, 1fr)`;
        }
    }};
grid-template-rows: ${props => {
        switch (props.size) {
            case 12:
                return `repeat(3, 1fr)`;
            case 20:
                return `repeat(4, 1fr)`;
            case 30:
                return `repeat(5, 1fr)`;
            default:
                return `repeat(4, 1fr)`;
        }
    }};
gap: 5px;
height: calc(.8 * 100vh);
width: 100vh;
margin-top: 1rem;

@media screen and (orientation: portrait){
height: calc(.8 * 95vw);
width: 95vw;
}
`

export const GameStats = styled('div') <{ gameOver: boolean }>`
position: absolute;
top: 10rem;
left: 50%;
transform: translateX(-50%);
display: flex;
height: 7rem;
width: 20rem;
align-items: center;
justify-content: space-between;
flex-direction: column;
font-size: 3rem;
color: white;
text-shadow: 1px 1px 5px rgba(0,0,0,.25);

p:last-child {
    font-size: 2rem;
}
`