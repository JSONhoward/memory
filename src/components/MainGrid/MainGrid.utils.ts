export const genBoxesArray = (num: number): Array<{col: number, row: number }> => {
    let arr: {col: number, row: number}[] = []
    let maxCol: number, maxRow: number
    if(num === 12) {
        maxCol = 4
        maxRow = 3
    }
    if(num === 20) {
        maxCol = 5
        maxRow = 4
    }
    if(num === 30) {
        maxCol = 6
        maxRow = 5
    }

    for (let i = 0; i < num; i++) {
        arr[i] = {col: i % maxCol! + 1, row: i % maxRow! + 1 }
    }

    return arr
}

export const sleep = (ms: number) => {
    return new Promise(resolve => setTimeout(resolve, ms))
}