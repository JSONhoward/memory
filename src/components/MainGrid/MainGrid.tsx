import React from 'react'

import GridBoxImage from './GridBoxImage/GridBoxImage'
import { genBoxesArray } from './MainGrid.utils';
import GridBox from './GridBox/GridBox';
import { GridObject, InitialTiles } from '../../@types/myTypes';
import { GameStats, StyledMainGrid } from './MainGrid.styles';

type Props = {
    handle: (event: any) => void,
    match: InitialTiles,
    pairNum: GridObject[],
    grid: GridObject[],
    show: InitialTiles,
    time: {start: number, end: number},
    won: boolean
}

const MainGrid: React.FC<Props> = ({ handle, match, pairNum, grid, show, time, won }) => {
    const boxes = genBoxesArray(grid.length).map((el, i) => {
        return (
            <GridBox match={match[`tile${i}` as keyof InitialTiles]} col={el.col} row={el.row} key={i}>
                <GridBoxImage name={`tile${i}`} pairNum={pairNum[i].pair} handle={handle} show={show[`tile${i}` as keyof InitialTiles]} image={grid[i].image} />
            </GridBox>
        )
    })

    return (
        <>
            <StyledMainGrid size={grid.length}>
                {boxes}
            </StyledMainGrid>
            {!won && (<GameStats gameOver={false}>
                <p>Game Over</p>
                <p>time: {((time.end - time.start) / 1000).toFixed(1)}s</p>
            </GameStats>)}
        </>
    )
}

export default MainGrid
