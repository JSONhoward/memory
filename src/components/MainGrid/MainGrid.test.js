import React from 'react'
import {create} from 'react-test-renderer'
import MainGrid from './'

const props = {
    handle: jest.fn(),
    match: [],
    pairNum: [],
    grid: [],
    show: [],
    time: {start: 1, end: 2},
    won: false
}

describe('MainGrid Tests', () => {
    test('Snapshot Test', () => {
        const tree = create(<MainGrid {...props} />).toJSON()

        expect(tree).toMatchSnapshot()
    })

    test('Receives Props', () => {
        const component = create(<MainGrid {...props} />).root

        expect(component.props).toEqual(props)
    })
})