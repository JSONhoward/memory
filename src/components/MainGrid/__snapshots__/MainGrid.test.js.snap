// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainGrid Tests Snapshot Test 1`] = `
<div
  className="sc-AxhCb cQWatt"
  size={0}
/>
`;
